File: doubly-linkedlist/doubly-linkedlist.go
/*
 Note. All functions are in-order

 ** Doubly Linked List
 ** New - new linkedlist{} and .initialize(value)
 ** Insert (at start) - .prepend(value)
 ** Insert (at end) - .append(value)
 ** Insert (at middle) - .insert(index, value)
 ** Lookup - .get(index)
 ** Remove - .remove(index)
*/

// =============================================

package main

import "fmt"

type node struct {
	value int
	next  *node
	prev  *node
}

type linkedlist struct {
	head   *node
	tail   *node
	length int
}

func (ll *linkedlist) initialize(value int) {
	n := node{value: value, next: nil, prev: nil}
	ll.head = &n
	ll.tail = &n
	ll.length = 1
}

func (ll linkedlist) isEmpty() bool {
	if ll.head == nil {
		fmt.Println("List is empty")
		return true
	}
	return false
}

func (ll *linkedlist) append(value int) {
	if ll.isEmpty() {
		ll.initialize(value)
		return
	}

	n := node{value: value, next: nil, prev: nil}
	ll.tail.next = &n
	n.prev = ll.tail
	ll.tail = &n
	ll.length++
}

func (ll *linkedlist) prepend(value int) {
	if ll.isEmpty() {
		ll.initialize(value)
		return
	}

	n := node{value: value, next: nil}
	n.next = ll.head
	ll.head.prev = &n
	ll.head = &n
	ll.length++
}

func (ll *linkedlist) print() {
	if !ll.isEmpty() {
		curr := ll.head
		nodes := []int{curr.value}
		curr = curr.next
		for curr != nil {
			nodes = append(nodes, (*curr).value)
			curr = curr.next
		}
		fmt.Println(nodes)
	}
}

func (ll *linkedlist) printBackwards() {
	if !ll.isEmpty() {
		curr := ll.tail
		nodes := []int{curr.value}
		curr = curr.prev
		for curr != nil {
			nodes = append(nodes, (*curr).value)
			curr = curr.prev
		}
		fmt.Println(nodes)
	}
}

func (ll *linkedlist) traverse(index int) (*node, *node, *node) {
	if ll.isEmpty() {
		return nil, nil, nil
	}

	curr := ll.head
	if index <= (ll.length / 2) {
		for i := 0; i < index && i < ll.length-1; i++ {
			curr = curr.next
		}
	} else {
		curr = ll.tail
		for i := ll.length - 1; i > index && i > 0; i-- {
			curr = curr.prev
		}
	}

	return curr.prev, curr, curr.next
}

func (ll *linkedlist) get(index int) (int, bool) {
	if !ll.isEmpty() {
		_, curr, _ := ll.traverse(index)
		return curr.value, true
	}

	return 0, false
}

func (ll *linkedlist) insert(index, value int) {
	if ll.isEmpty() {
		ll.initialize(value)
		return
	}
	n := node{value: value, next: nil}

	_, curr, leader := ll.traverse(index)
	curr.next = &n
	n.prev = curr
	n.next = leader
	if leader != nil {
		leader.prev = &n
	} else {
		ll.tail = &n
	}
	ll.length++
}

func (ll *linkedlist) remove(index int) {
	if !ll.isEmpty() {
		follower, curr, leader := ll.traverse(index)
		if follower == nil {
			ll.head = leader
			leader.prev = nil
			curr.next = nil
		} else if leader == nil {
			ll.tail = follower
			follower.next = nil
			curr.prev = nil
		} else {
			follower.next = leader
			leader.prev = follower
			curr.next = nil
			curr.prev = nil
		}
		ll.length--
	}
}

func main() {
	ll := linkedlist{}
	ll.print()
	ll.initialize(5)
	ll.append(16)
	ll.prepend(10)
	ll.print()
	ll.printBackwards()
}

File: doubly-linkedlist/doubly-linkedlist.js
/*
 Note. All functions are in-order

 ** Doubly Linked List 
 ** New - new LinkedList(value)
 ** Insert (at start) - .prepend(value)
 ** Insert (at end) - .append(value)
 ** Insert (at middle) - .insert(index, value)
 ** Lookup - .get(index)
 ** Remove - .remove(index)
 */

// =============================================

class LinkedList {
  constructor(value) {
    this.initialize(value);
  }

  initialize = (value) => {
    this.head = { value, next: null, prev: null };
    this.tail = this.head;
    this.length = 1;
  };

  isEmpty = () => {
    if (!this.head) {
      console.log('List is empty');
      return true;
    }
    return false;
  };

  append = (value) => {
    if (this.isEmpty()) {
      this.initialize(value);
      return this;
    }

    const node = { value, next: null, prev: null };
    this.tail.next = node;
    node.prev = this.tail;
    this.tail = node;
    this.length++;
    return this;
  };

  prepend = (value) => {
    if (this.isEmpty()) {
      this.initialize(value);
      return this;
    }

    const node = { value, next: null, prev: null };
    node.next = this.head;
    this.head.prev = node;
    this.head = node;
    this.length++;
    return this;
  };

  print = () => {
    if (!this.isEmpty()) {
      let curr = this.head;
      let str = curr.value + '';
      while (curr.next != null) {
        curr = curr.next;
        str += ` <-> ${curr.value}`;
      }
      console.log(str);
    }
  };

  printBackwards = () => {
    if (!this.isEmpty()) {
      let curr = this.tail;
      let str = curr.value + '';
      while (curr.prev != null) {
        curr = curr.prev;
        str += ` <-> ${curr.value}`;
      }
      console.log(str);
    }
  };

  traverse = (index) => {
    if (this.isEmpty()) {
      return { follower: null, curr: null, leader: null };
    }

    let curr = this.head;
    if (index <= this.length / 2) {
      for (let i = 0; i < index && i < this.length - 1; i++) {
        curr = curr.next;
      }
    } else {
      curr = this.tail;
      for (let i = this.length - 1; i > index && i > 0; i--) {
        curr = curr.prev;
      }
    }

    return { follower: curr.prev, curr, leader: curr.next };
  };

  get = (index) => {
    if (!this.isEmpty()) {
      return this.traverse(index).curr.value;
    }
    return null;
  };

  insert = (index, value) => {
    if (this.isEmpty()) {
      this.initialize(value);
      return this;
    }
    const node = { value, next: null };

    const { _, curr, leader } = this.traverse(index);
    curr.next = node;
    node.prev = curr;
    node.next = leader;
    if (leader) {
      leader.prev = node;
    } else {
      this.tail = node;
    }
    this.length++;
    return this;
  };

  remove = (index) => {
    if (!this.isEmpty()) {
      const { follower, curr, leader } = this.traverse(index);
      if (!follower) {
        this.head = leader;
        leader.prev = null;
        curr.next = null;
      } else if (!leader) {
        this.tail = follower;
        follower.next = null;
        curr.prev = null;
      } else {
        follower.next = leader;
        leader.prev = follower;
        curr.next = null;
        curr.prev = null;
      }
      this.length--;
    }

    return this;
  };
}

const ll = new LinkedList(5);
ll.print();
ll.printBackwards();
ll.append(15);
ll.print();
ll.printBackwards();
ll.prepend(11);
ll.print();
ll.printBackwards();
